feat: accounts progress; use bun for the api server;

lmssiehdev opened the pull request:
## Endpoints 

- [x] `/api/profile/`
  - [ ] loadoutPriv 
- [x] `/api/username/`
  - [x] Username change time
- [x] `/api/loadout/`
- [x] `/api/logout/`
- [x] `/api/delete/`
- [x] `/api/auth/`
  - [x] Google
  - [x] Discord
- [x] `/api/set_item_status/`
- [x] `/api/unlock/`

## Minimal guide

Accounts are optional, set `accountsEnabled` to `false` in `config.ts` to disable them.
If disabled, you can skip these steps.

```bash
 cd /server

 # Copy the example environment file
 # optional during development, you can test account features without it
 cp .env.example .env

 # Generate and apply the database schema (creates and populates the SQLite file)
 # run this everytime you make changes to the schema.ts
 pnpm run db:generate
 pnpm run db:migrate

 # Start the server
 pnpm run dev
 # or
 # pnpm run dev:api
 # pnpm run dev:game
```

worth skimming over https://orm.drizzle.team/docs/migrations#migrations

Jarred-Sumner commented:
nice

leia-uwu commented:
looking good!
will def merge when its ready!

lmssiehdev commented:
ignoring the conflicts, this should be ready for review :crossed_fingers:

leia-uwu commented on client/src/ui/loadouts.ts:
```diff
@@ -127,13 +44,15 @@ const loadout = {
-    getUserAvailableItems: function (heroItems: unknown[]) {
-        const items = [];
+    getUserAvailableItems: function (heroItems: Item[] = []) {
+        const processedItemTypes = new Set<string>();
```
why is this set necessary
shouldn't duplicated items validation be done server side?

leia-uwu commented:
ngl i kinda of don't want github login because like at max 3 people will use it knowing our player base
feels just like more code to maintain
i would go with google and discord login only

lmssiehdev commented:
GitHub was the easiest to get started with, which is why I added it. Now that I've added a "mock" auth for testing during development there is no reason to keep it.

I did merge the feat/stats branch into this and got a working user stats and leaderboard version working, but I'll create a separate PR for that later I think after I get enough feedback on this.

DamienVesper commented:
Don't like that `arctic` doesn't follow semver, but that's just my nitpick.

leia-uwu commented:
> GitHub was the easiest to get started with, which is why I added it. Now that I've added a "mock" auth for testing during development there is no reason to keep it.
> 
> I did merge the feat/stats branch into this and got a working user stats and leaderboard version working, but I'll create a separate PR for that later I think after I get enough feedback on this.

nice!
can i solve the conflicts for you tomorrow if you didn't already locally? I'm interested in merging this in like 2 weeks or so, probably after halloween releases

leia-uwu commented:
wait nvm still need google login i forgot

DamienVesper commented:
google is easy, probably resolve and merge in new commit later, takes like 5 min to write and test

lmssiehdev commented:
I don't think I can get to the conflicts today, so feel free to resolve them.

leia-uwu commented:
rebasing rn
damn bun / hono HTTP api is so much fucking better than uws

leia-uwu commented:
i think we should be consistent about api keys on config or env
i like env more even if it ads fragmentation
but with the config json thats on gitignore being merged with the config so you have a way to not push local configs idk if theres a reason to use env

lmssiehdev commented:
I think we should keep all secrets in the .env file and the config for game related stuff, but whatever option is fine honestly.

leia-uwu commented:
ye i think .env is the best

leia-uwu commented:
are the files on server/src/api/db/drizzle supposed to be committed?

lmssiehdev commented:
It's recommended, yes.
